Add Event Types

greggman opened the pull request:
this makes `addEventListener` on `GPUDevice` have the right type information.

![Screenshot 2025-01-29 at 11 50 18](https://github.com/user-attachments/assets/f057ab5e-23ff-4661-b1d1-6e30f102a6fe)

![Screenshot 2025-01-29 at 11 51 12](https://github.com/user-attachments/assets/14684566-34d4-4396-9071-4aeb59c5ec47)

I wasn't sure how to add it to the generated file.

greggman commented:
maybe these should go in the spec so they're upstream?

kainino0x commented on dist/main.d.ts:
```diff
@@ -0,0 +1,16 @@
+import './index.d.ts';
```
Historically we've been just adding things like this to the top of `index.d.ts`:
https://github.com/gpuweb/types/blob/8faeb6d6df4d6f577a34a65b3bf768b8bd2714cd/dist/index.d.ts#L5-L49
It's probably fine to have multiple files, but we should have all of the additions in one place.

greggman replied:
moved what I could.

kainino0x commented on dist/main.d.ts:
```diff
@@ -0,0 +1,16 @@
+import './index.d.ts';
```
nit: I don't think we should have "index" and "main" as those both denote main files. We could move the existing `index.d.ts` to another filename though.

greggman replied:
moved it to `types.d.ts`

kainino0x reviewed (commented):
> maybe these should go in the spec so they're upstream?

There is unfortunately no way to describe these in WebIDL (except for `attribute EventHandler onuncapturederror;`). 

I actually thought we had this already, but I guess not!

greggman commented:
note: I separated this into 2 comments. If you look at the commit diffs instead of the PR diff it's easier to see the changes.

greggman commented:
sorry, I'll fix the tests

greggman commented:
ugh. Ok. I just added the types to the original `dist/index.d.ts`

I could not figure out how to separate it into 2 files.

The issue was with `index.d.ts` (new file with manual types) importing `types.d.ts` (the IDL generated file). It was getting errors about types it didn't know. See https://github.com/gpuweb/types/actions/runs/13063705687/job/36452157799#step:5:18

So, I imported those with

```
import {
  GPUCanvasConfigure,
  GPUIntegerCoordinate',
  ...
} "./types.d.ts";
```

This shut-up TS about undefined types but now it complained `types.d.ts` is not a module (some voodoo magic ts stuff). If I make it a module (needs to export something) then it would no longer add its types to the gloal since it's now a module and not some type file. To make it export its types requires wrapping them all in `declare global { ... }` but if I do that then prettier is going to reformat the entire file with 2 space indents and that means the diff between `generated/index.d.ts` and `dist/types.d.ts` would be every line. I could add some code to generate `generated/index.d.ts` with those lines and that formatting but it was getting ridiculous. My goal was to add ~7 lines of code. Not to spend hours making TS happy.

So, if we want those separate files we can do that in another PR. For now I just added the changes to `dist/index.d.ts`

kainino0x commented on dist/index.d.ts:
```diff
@@ -2447,6 +2447,10 @@ declare var GPUComputePipeline: {
   new (): never;
 };
 
+interface GPUDeviceEventMap {
```
nit: `/** @internal */`, and maybe prefix with `__` or something like we do for `__brand`

greggman replied:
done

kainino0x reviewed (approved):
LGTM

kainino0x commented:
Thanks for looking into splitting it into two files. It's not super important, so if we can't easily figure out how to do it then it's fine if we don't. It also matters less until we have the codegen working better (assuming we ever bother to fix it).

kainino0x commented:
Just remembered to publish: 0.1.54
